File: TendWeb/src/lib/eventStore.ts
/**
 * The store-facing side of recurrence.
 *
 * It decides nothing: `recurrence.ts` works out what an edit means and this
 * applies the result to the events array in one pass, so a half-applied split
 * can't exist.
 */

import {
  applyDelete,
  applyEdit,
  applySnapshot,
  formatRule,
  mergeOccurrences,
  occurrences,
  toSnapshot,
  type EventEdit,
  type Occurrence,
  type SeriesMutation,
  type Snapshot,
} from './recurrence'
import type { EditScope, TendEvent } from '../types'

export interface ExpandedEvent {
  event: TendEvent
  start: Date
  end: Date
  isDetached: boolean
  /** The occurrence the user is looking at — what an edit must be anchored to. */
  occurrenceStart: Date
}

/** Every occurrence visible in a range, series expanded and detached copies folded in. */
export function expandEvents(events: TendEvent[], from: Date, to: Date): ExpandedEvent[] {
  const snapshots = events.map(toSnapshot)
  const byId = new Map(events.map((e) => [e.id, e]))

  const detachedBySeries = new Map<string, Snapshot[]>()
  for (const snapshot of snapshots) {
    if (!snapshot.detachedFrom) continue
    const list = detachedBySeries.get(snapshot.seriesId) ?? []
    list.push(snapshot)
    detachedBySeries.set(snapshot.seriesId, list)
  }

  const results: Occurrence[] = []
  for (const snapshot of snapshots) {
    if (snapshot.detachedFrom) continue
    const expanded = occurrences(snapshot, from, to)
    results.push(...mergeOccurrences(expanded, detachedBySeries.get(snapshot.seriesId) ?? [], from, to))
  }

  return results
    .sort((a, b) => a.start.getTime() - b.start.getTime())
    .flatMap((occurrence) => {
      const event = byId.get(occurrence.eventId)
      if (!event) return []
      return [
        {
          event,
          start: occurrence.start,
          end: occurrence.end,
          isDetached: occurrence.isDetached,
          occurrenceStart: occurrence.start,
        },
      ]
    })
}

/** The rule-bearing event for a series. A detached copy has to reach back to it. */
function masterFor(events: TendEvent[], event: TendEvent): TendEvent {
  if (event.recurrenceRule && !event.detachedFrom) return event
  return (
    events.find((e) => e.seriesId === event.seriesId && e.recurrenceRule && !e.detachedFrom) ?? event
  )
}

function detachedFor(events: TendEvent[], seriesId: string): TendEvent[] {
  return events.filter((e) => e.seriesId === seriesId && e.detachedFrom)
}

function applyMutation(
  events: TendEvent[],
  master: TendEvent,
  mutation: SeriesMutation,
  by: string | undefined,
): TendEvent[] {
  const deleted = new Set(mutation.deletedEventIds)
  const now = Date.now()

  let next = events
    .filter((e) => !deleted.has(e.id))
    .map((e) => {
      if (e.id === master.id && !deleted.has(e.id)) {
        return applySnapshot(e, mutation.updatedMaster, by)
      }
      const reassigned = mutation.reassignedSeries[e.id]
      if (reassigned) return { ...e, seriesId: reassigned, updatedAt: now }
      return e
    })

  for (const created of mutation.createdEvents) {
    next = [
      ...next,
      applySnapshot(
        {
          id: created.id,
          title: created.title,
          start: created.start.toISOString(),
          end: created.end.toISOString(),
          isAllDay: created.isAllDay,
          seriesId: created.seriesId,
          exceptionDates: [],
          attendeeIds: created.attendeeIds,
          createdAt: now,
        },
        created,
        by,
      ),
    ]
  }

  return next
}

/**
 * Applies an edit under an explicit scope.
 *
 * `occurrenceStart` is the instance the user was looking at, not the series
 * start. Passing the series start for a "this and future" edit is the classic
 * bug this signature is shaped to prevent.
 */
export function editEvent(
  events: TendEvent[],
  event: TendEvent,
  edit: EventEdit,
  scope: EditScope,
  occurrenceStart: Date,
  by: string | undefined,
): TendEvent[] {
  const master = masterFor(events, event)
  const detached = detachedFor(events, master.seriesId).map(toSnapshot)
  const mutation = applyEdit(toSnapshot(master), edit, scope, occurrenceStart, detached)
  return applyMutation(events, master, mutation, by)
}

export function deleteEvent(
  events: TendEvent[],
  event: TendEvent,
  scope: EditScope,
  occurrenceStart: Date,
  by: string | undefined,
): TendEvent[] {
  const master = masterFor(events, event)
  const detached = detachedFor(events, master.seriesId).map(toSnapshot)
  const mutation = applyDelete(toSnapshot(master), scope, occurrenceStart, detached)
  return applyMutation(events, master, mutation, by)
}

export function makeEvent(fields: {
  title: string
  start: Date
  end: Date
  isAllDay?: boolean
  location?: string
  notes?: string
  attendeeIds?: string[]
  rule?: ReturnType<typeof formatRule> | null
  by?: string
}): TendEvent {
  const id = crypto.randomUUID()
  return {
    id,
    title: fields.title,
    location: fields.location,
    start: fields.start.toISOString(),
    end: fields.end.toISOString(),
    isAllDay: fields.isAllDay ?? false,
    recurrenceRule: fields.rule ?? undefined,
    // A one-off is its own series of one, so every event has a series to belong
    // to and nothing has to special-case `null` later.
    seriesId: id,
    exceptionDates: [],
    attendeeIds: fields.attendeeIds ?? [],
    notes: fields.notes,
    lastModifiedBy: fields.by,
    createdAt: Date.now(),
    updatedAt: Date.now(),
  }
}

File: Tend/TendKit/Sources/TendKit/Services/EventStore.swift
import Foundation
import SwiftData

public extension Event {
    var snapshot: EventSnapshot {
        EventSnapshot(
            id: id,
            seriesID: seriesID ?? id,
            title: title,
            location: location,
            startDate: startDate,
            endDate: endDate,
            isAllDay: isAllDay,
            rule: recurrenceRule.flatMap(RecurrenceRule.init(rawValue:)),
            exceptionDates: exceptionDates,
            detachedFromOccurrence: detachedFromOccurrence,
            attendeeIDs: attendeeIDs,
            notes: notes
        )
    }

    func apply(_ snapshot: EventSnapshot, by memberID: UUID?, at date: Date = .now) {
        title = snapshot.title
        location = snapshot.location
        startDate = snapshot.startDate
        endDate = snapshot.endDate
        isAllDay = snapshot.isAllDay
        recurrenceRule = snapshot.rule?.rawValue
        seriesID = snapshot.seriesID
        exceptionDates = snapshot.exceptionDates
        detachedFromOccurrence = snapshot.detachedFromOccurrence
        attendeeIDs = snapshot.attendeeIDs
        notes = snapshot.notes
        lastModifiedBy = memberID
        lastModifiedAt = date
    }

    static func make(from snapshot: EventSnapshot, by memberID: UUID?) -> Event {
        let event = Event(
            id: snapshot.id,
            title: snapshot.title,
            location: snapshot.location,
            startDate: snapshot.startDate,
            endDate: snapshot.endDate,
            isAllDay: snapshot.isAllDay,
            recurrenceRule: snapshot.rule?.rawValue,
            seriesID: snapshot.seriesID,
            attendeeIDs: snapshot.attendeeIDs,
            notes: snapshot.notes,
            lastModifiedBy: memberID
        )
        event.exceptionDates = snapshot.exceptionDates
        event.detachedFromOccurrence = snapshot.detachedFromOccurrence
        return event
    }
}

/// The SwiftData side of recurrence. It decides *nothing* — ``RecurrenceEngine``
/// works out what an edit means and this applies the result in one save, so a
/// half-applied split can't exist.
public enum EventStore {

    // MARK: - Reading

    /// Every occurrence visible in a date range, series expanded and detached
    /// copies folded in.
    public static func occurrences(
        in range: Range<Date>,
        context: ModelContext,
        calendar: Calendar = .current
    ) -> [EventOccurrence] {
        let events = (try? context.fetch(FetchDescriptor<Event>())) ?? []
        return occurrences(in: range, events: events, calendar: calendar)
    }

    public static func occurrences(
        in range: Range<Date>,
        events: [Event],
        calendar: Calendar = .current
    ) -> [EventOccurrence] {
        let snapshots = events.map(\.snapshot)
        let detachedBySeries = Dictionary(grouping: snapshots.filter { $0.detachedFromOccurrence != nil }) {
            $0.seriesID
        }

        var results: [EventOccurrence] = []
        for snapshot in snapshots where snapshot.detachedFromOccurrence == nil {
            let expanded = RecurrenceEngine.occurrences(of: snapshot, in: range, calendar: calendar)
            results.append(
                contentsOf: RecurrenceEngine.merge(
                    occurrences: expanded,
                    detached: detachedBySeries[snapshot.seriesID] ?? [],
                    in: range
                )
            )
        }
        return results.sorted { $0.start < $1.start }
    }

    public static func event(id: UUID, context: ModelContext) -> Event? {
        var descriptor = FetchDescriptor<Event>(predicate: #Predicate { $0.id == id })
        descriptor.fetchLimit = 1
        return (try? context.fetch(descriptor))?.first
    }

    /// True when the editor has to ask "this event / this and future / all".
    /// A one-off event never prompts, and neither does a series with exactly
    /// one remaining occurrence.
    public static func requiresScopePrompt(for event: Event) -> Bool {
        guard let rule = RecurrenceRule(rawValue: event.recurrenceRule ?? "") else { return false }
        if let count = rule.count, count <= 1 { return false }
        return true
    }

    // MARK: - Writing

    /// Applies an edit under an explicit scope.
    ///
    /// `occurrenceStart` is the instance the user was actually looking at, not
    /// the series start. Passing the series start for a "this and future" edit
    /// is the classic bug this signature is shaped to prevent.
    @discardableResult
    public static func applyEdit(
        _ edit: EventEdit,
        scope: RecurrenceEditScope,
        to event: Event,
        occurrenceStart: Date,
        by memberID: UUID?,
        context: ModelContext,
        calendar: Calendar = .current
    ) throws -> SeriesMutation {
        let master = masterEvent(for: event, context: context) ?? event
        let detached = detachedCopies(ofSeries: master.seriesID ?? master.id, context: context)

        let mutation = RecurrenceEngine.apply(
            edit: edit,
            scope: scope,
            to: master.snapshot,
            occurrenceStart: occurrenceStart,
            detachedCopies: detached.map(\.snapshot),
            calendar: calendar
        )

        master.apply(mutation.updatedMaster, by: memberID)
        master.recurrenceEditScope = mutation.appliedScope.rawValue

        for snapshot in mutation.createdEvents {
            let created = Event.make(from: snapshot, by: memberID)
            created.recurrenceEditScope = mutation.appliedScope.rawValue
            context.insert(created)
        }

        for (eventID, newSeriesID) in mutation.reassignedSeries {
            detached.first { $0.id == eventID }?.seriesID = newSeriesID
        }

        for eventID in mutation.deletedEventIDs {
            if let doomed = self.event(id: eventID, context: context) { context.delete(doomed) }
        }

        try context.save()
        return mutation
    }

    @discardableResult
    public static func delete(
        _ event: Event,
        scope: RecurrenceEditScope,
        occurrenceStart: Date,
        by memberID: UUID?,
        context: ModelContext,
        calendar: Calendar = .current
    ) throws -> SeriesMutation {
        let master = masterEvent(for: event, context: context) ?? event
        let detached = detachedCopies(ofSeries: master.seriesID ?? master.id, context: context)

        let mutation = RecurrenceEngine.applyDelete(
            scope: scope,
            to: master.snapshot,
            occurrenceStart: occurrenceStart,
            detachedCopies: detached.map(\.snapshot),
            calendar: calendar
        )

        if mutation.deletedEventIDs.contains(master.id) {
            for eventID in mutation.deletedEventIDs {
                if let doomed = self.event(id: eventID, context: context) { context.delete(doomed) }
            }
        } else {
            master.apply(mutation.updatedMaster, by: memberID)
            for eventID in mutation.deletedEventIDs {
                if let doomed = self.event(id: eventID, context: context) { context.delete(doomed) }
            }
        }

        try context.save()
        return mutation
    }

    // MARK: - Series helpers

    /// The rule-bearing event for a series. A detached copy has no rule, so
    /// editing one with "all events" has to reach back to the master.
    static func masterEvent(for event: Event, context: ModelContext) -> Event? {
        guard let seriesID = event.seriesID else { return event }
        if event.recurrenceRule != nil, event.detachedFromOccurrence == nil { return event }
        let descriptor = FetchDescriptor<Event>(predicate: #Predicate { $0.seriesID == seriesID })
        return (try? context.fetch(descriptor))?.first { $0.recurrenceRule != nil && $0.detachedFromOccurrence == nil }
    }

    static func detachedCopies(ofSeries seriesID: UUID, context: ModelContext) -> [Event] {
        let descriptor = FetchDescriptor<Event>(predicate: #Predicate { $0.seriesID == seriesID })
        return ((try? context.fetch(descriptor)) ?? []).filter { $0.detachedFromOccurrence != nil }
    }
}
